qip-0012.md

```diff
@@ -0,0 +1,51 @@
+```
+ QIP: 12
+ Layer: Consensus (hard fork)
+ Title: Qi UTXO Pruning
+ Author: wizeguyy <wizeguyy+qip@quai.org>
+ Comments-Summary: No comments yet.
+ Comments-URI: https://github.com/quainetwork/qips/wiki/Comments:QIP-0012
+ Status: Draft
+ Type: Standards Track
+ Created: 2024-04-02
+ License: BSD-2-Clause
+```
+
+## Abstract
+This QIP proposes a mechanism to prune dust UTXOs from the ledger.
+
+## Motivation
+Transaction outputs in the Qi UTXO ledger, occupy space and impose a burden on
+UTXO transaction processing, from the set root recomputations. This is not a
+problem under normal usage, but can become a problem if any users/wallets are
+lazy or irresponsible with UTXO reconsolidation, or worse if the network
+undergoes a spam attack. This QIP describes a protocol change to disincentivize
+UTXO state pollution without meaningfully impacting normal transactional
+behavior.
+
+## Rationale
+Normal Qi wallet behavior will involve some amount of UTXO splitting, to create
+change outputs for payments, as well as UTXO reconsolidation as the wallet
+acquires too many small change outputs. This QIP should not negatively impact
+this expected user behavior, but should provide a cost to lazy wallets that do
+not reconsolidate change outputs.
+
+## Specification
+We achieve this simply, by limiting the size of the UTXO set. If a transaction
+creates new UTXOs which would exceed the UTXO set capacity, we destroy the
+smallest UTXOs in the ledger. This is practical to do thanks to the fixed
+denominations in the Qi ledger.
+
+Since the network cost we are trying to control is related to the construction
+of the UTXO set root, it is sensible to choose a capacity that reflects a
+particular depth of the state tree. Since we use a Patricia Merkle trie, the
+state root recomputation is bounded by the radix-16 update cost bound:
+$O(log_{16}(N))$, where $N$ is the number of UTXOs in the set.
+
+We set a max trie depth size of 10, which corresponds to a max UTXO set size of
+$16^10 \appox ~1 trillion$ UTXOs. If a transaction mints new UTXOs which exceed
+the $16^10$ limit, the node shall prune all of the smallest UTXOs from the UTXO
```

Comment:
How long would it take to recompute the root of the trie with 1 trillion nodes/depth of 10? Average case for say an 8-core CPU?

Reply:
Well, the computation isnt the limit. Its usually the disk IOPS to read write each trie node that limits you. To strictly answer you question, here's some napkin math:

An 8c/16t 4GHz CPU, assuming keccak takes 15 cycles per byte (source wikipedia) and each trie node being up to 17*32 bytes:
```
17 * 32 * 15 = 8160 cycles / trie node hash
4GHz / 8160 = 490K tries/s per thread
490K tries/s * 16 threads = 7.8M tries/s
10 tries / 7.8M tries/s ~= 1.36us / root calculation
```

But, as I mentioned at the start, this is just the compute component. The dominant cost is actually the IOPS the disk can handle. A high end SSD tends to get around 45K IOPS, which equates to ~= 23us per disk access. At 10 trie levels, you need 460 us just to read 10 original nodes and write 10 new nodes, as well as 2x23us for the leaf node itself. Lets call it 500us to add a single UTXO to the trie. So, we could add 2K UTXOs per second to the trie at level 10.

That is the naive implementation. A good implementation will amortize some of those costs with batch operations, but that's beyond the scope of my napkin math. There's also some costs not accounted for here, e.g. time to look up and remove spent UTXOs from the set.

Reply:
Where did `10 tries` come from? I'm curious about the length of time to recompute the root of a PMT with a trillion elements. Shouldn't the computation use `1 trillion tries`

Reply:
I assume you mean 1 trillion leafs/accounts, not including all the intermediate trie nodes, right?

10 trie nodes is the number of levels in $log_{16}$ for a trie with 1 trillion leafs. An optimized PMT update algorithm (not recomputing from scratch) can update the root in $log_{16}$.

I thought you were asking for the CPU time, so I gave you some napkin math for that, but I realize now you are just asking for total recomputation time, which again is dominated by IOPS, not CPU performance.

There are a LOT of factors that could influence disk access performance (disk speed, how busy is the disk with other software?, database in-memory caching / paging strategies, etc), so its not reasonable to try and "napkin math" it here. You'd have to benchmark a particular implementation to get an idea.